osu.Game/Skinning/SkinProvidingContainer.cs

```diff
@@ -53,163 +40,179 @@ public class SkinProvidingContainer : Container, ISkinSource
             {
                 ISample sourceSample;
-                if ((sourceSample = disableableSkinSources[skin]?.GetSample(sampleInfo)) != null)
+                if ((sourceSample = lookupWrapper.GetSample(sampleInfo)) != null)
                     return sourceSample;
             }
 
-            return fallbackSource?.GetSample(sampleInfo);
+            if (!AllowFallingBackToParent)
+                return null;
+
+            return ParentSource?.GetSample(sampleInfo);
         }
 
         public IBindable<TValue> GetConfig<TLookup, TValue>(TLookup lookup)
         {
-            foreach (var skin in SkinSources)
+            foreach (var (_, lookupWrapper) in skinSources)
             {
                 IBindable<TValue> bindable;
-                if ((bindable = disableableSkinSources[skin]?.GetConfig<TLookup, TValue>(lookup)) != null)
+                if ((bindable = lookupWrapper.GetConfig<TLookup, TValue>(lookup)) != null)
                     return bindable;
             }
 
-            return fallbackSource?.GetConfig<TLookup, TValue>(lookup);
+            if (!AllowFallingBackToParent)
+                return null;
+
+            return ParentSource?.GetConfig<TLookup, TValue>(lookup);
         }
 
-        protected virtual void OnSourceChanged() => SourceChanged?.Invoke();
+        /// <summary>
+        /// Add a new skin to this provider. Will be added to the end of the lookup order precedence.
+        /// </summary>
+        /// <param name="skin">The skin to add.</param>
+        protected void AddSource(ISkin skin)
+        {
+            skinSources.Add((skin, new DisableableSkinSource(skin, this)));
+
+            if (skin is ISkinSource source)
+                source.SourceChanged += TriggerSourceChanged;
+        }
 
-        protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnlyDependencyContainer parent)
+        /// <summary>
+        /// Remove a skin from this provider.
+        /// </summary>
+        /// <param name="skin">The skin to remove.</param>
+        protected void RemoveSource(ISkin skin)
         {
-            var dependencies = new DependencyContainer(base.CreateChildDependencies(parent));
+            if (skinSources.RemoveAll(s => s.skin == skin) == 0)
+                return;
 
-            if (AllowFallingBackToParent)
-            {
-                fallbackSource = dependencies.Get<ISkinSource>();
-                if (fallbackSource != null)
-                    fallbackSource.SourceChanged += OnSourceChanged;
-            }
+            if (skin is ISkinSource source)
+                source.SourceChanged -= TriggerSourceChanged;
+        }
 
-            dependencies.CacheAs<ISkinSource>(this);
+        /// <summary>
+        /// Clears all skin sources.
+        /// </summary>
+        protected void ResetSources()
+        {
+            foreach (var i in skinSources)
+                RemoveSource(i.skin);
```

Comment:
This is an error-prone loop. Maybe `.ToArray()` it.

Reply:
oops.